### PR TITLE
Add GenericType on Expo Notifications

### PR DESCRIPTION
**Pull request by victorwads**
Add workaround for issue: https://github.com/expo/expo/issues/10070

# Why
Add GenericType for typing notification data, this way it helps on CodeCompleting on data received on push notifications.
And of course, organizing your data by typing everything =D.

# How
Provides a TypscriptType typed interfaces for developer get the wanted data:

# Test Plan
https://forums.expo.io/t/expo-notification-receivedlistener-wrong-data-on-ios/42469/3
```typescript
interface ScheduleNotificationData {
    scheduleID: string
    userId: string
}
const notification = addNotificationReceivedListener<ScheduleNotificationData>(received => {
  load(true)
  console.warn('addNotificationReceivedListener', received)
})
```
## Also work without explicitly Typing
```typescript
const notification = addNotificationReceivedListener(received => {
  load(true)
  console.warn('addNotificationReceivedListener', received)
})
```

### CODE REVIEW

**Comment by victorwads** [comment]
Hi @cruzach,
Hi @sjchmiela.

I don't understand why the first job failed, the second job looks like some CI config.
Should I do something?

File: packages/expo-notifications/src/Notifications.types.ts
Comment (lukmccall) [inline packages/expo-notifications/src/Notifications.types.ts:186]:
```suggestion
export type NotificationContent<T extends Record<string, unknown> = Record<string, unknown>>
```
I think we should constrain T to be only the `Record` type.

File: packages/expo-notifications/src/Notifications.types.ts
Comment (lukmccall) [inline packages/expo-notifications/src/Notifications.types.ts:192]:
```suggestion
  data: T;
```
This should be enough here, cause you passed the default type for `T`, which is `Record<string, unknow>.

File: packages/expo-notifications/src/Notifications.types.ts
Comment (lukmccall) [inline packages/expo-notifications/src/Notifications.types.ts:220]:
```suggestion
export interface NotificationRequest<T extends Record<string, unknown> = Record<string, unknown>> {
```
I'm not a fan of the `unknown` type :/

File: packages/expo-notifications/src/Notifications.types.ts
Comment (lukmccall) [inline packages/expo-notifications/src/Notifications.types.ts:261]:
```suggestion
export interface Notification<T extends Record<string, unknown> = Record<string, unknown>> {
```
Same here

File: packages/expo-notifications/src/Notifications.types.ts
Comment (lukmccall) [inline packages/expo-notifications/src/Notifications.types.ts:266]:
```suggestion
export interface NotificationResponse<T extends Record<string, unknown> = Record<string, unknown>> {
```
Same here

File: packages/expo-notifications/src/NotificationsEmitter.ts
Comment (lukmccall) [inline packages/expo-notifications/src/NotificationsEmitter.ts:15]:
```suggestion
export function addNotificationReceivedListener<T extends Record<string, unknown> = Record<string, unknown>> {
```
Same here

File: packages/expo-notifications/src/NotificationsEmitter.ts
Comment (lukmccall) [inline packages/expo-notifications/src/NotificationsEmitter.ts:25]:
```suggestion
export function addNotificationResponseReceivedListener<T extends Record<string, unknown> = Record<string, unknown>>(
```
and here ;)

**Review by lukmccall — CHANGES_REQUESTED** [review]
LGTM 🎉
I've left some comments. 
And I'm not sure if `addNotificationReceivedListener` and `addNotificationResponseReceivedListener` are the only functions which use changed types.

File: packages/expo-notifications/src/Notifications.types.ts
Comment (victorwads) [reply packages/expo-notifications/src/Notifications.types.ts:186]: I'll submit this changes today!

File: packages/expo-notifications/src/Notifications.types.ts
Comment (victorwads) [reply packages/expo-notifications/src/Notifications.types.ts:192]: I'll submit these changes today!

File: packages/expo-notifications/src/Notifications.types.ts
Comment (victorwads) [reply packages/expo-notifications/src/Notifications.types.ts:261]: I'll submit these changes today!

File: packages/expo-notifications/src/Notifications.types.ts
Comment (victorwads) [reply packages/expo-notifications/src/Notifications.types.ts:266]: I'll submit these changes today!

File: packages/expo-notifications/src/NotificationsEmitter.ts
Comment (victorwads) [reply packages/expo-notifications/src/NotificationsEmitter.ts:15]: I'll submit these changes today!

File: packages/expo-notifications/src/NotificationsEmitter.ts
Comment (victorwads) [reply packages/expo-notifications/src/NotificationsEmitter.ts:25]: I'll submit these changes today!

**Comment by victorwads** [comment]
@lukmccall Done!, Thanks

**Review by lukmccall — APPROVED** [review]
@victorwads, thanks 🎉  I'm waiting for @sjchmiela's approval, and then I'll merge it ;)

**Comment by sjchmiela** [comment]
Hey all,

after some thought I don't think landing this pull request in the current state would provide the best developer experience possible (and the one which we all strive to provide).

At first this looks like a simple enough change which makes it easy for developers to type notification `data` and write type-safe code, but after taking edge cases into account things may go wrong and become more difficult to debug than without this change.

**tl;dr** — We can't trust notification data will conform to some type without checking first and API. If a developer feels safe about such change they can use `as` TS keyword. If they want to not only feel safe, but also stay safe they can use some kind of [validator](https://github.com/vriad/zod) that will ensure that the data conforms to some type.

We have an idea how to improve this notifications API endpoint even further to make it super-easy to write cross-platform notification listening code, stay tuned for more information!

---

Imagine this change is introduced, a developer adds a type to the received `data` field:
```ts
type MyNotificationData = {
  count: number;
}
```
Great! Let's say the developer also adds a notification listener that updates some stored information:
```ts
import * as Notifications from "expo-notifications";
import { AsyncStorage } from "react-native";

Notifications.addNotificationReceivedListener<MyNotificationData>(notification => {
  AsyncStorage.setItem("LAST_POWER", Math.pow(notification.content.data.count, 2));
});
```
Everything looks safe, `notification.content.data` renders as `MyNotificationData` which, according to type definition, always has a property `count` of type `number`.

However the Notifications API can't really always ensure that the `data` is of this particular type. Imagine a new feature is added to the app the developer builds which makes it receive more notifications of different types, eg.

```ts
type NewNotificationData = {
  name: string;
};

Notifications.addNotificationReceivedListener<MyNotificationData | NewNotificationData>(notification => {
  if (notification.content.data.count) {
    AsyncStorage.setItem("LAST_POWER", Math.pow(notification.content.data.count, 2));
  } // ...
});
```

Whenever apps with only the old implementation of the listener receive a new notification (and it's difficult to keep track of what version is currently installed on which device etc.) the app could crash! (`Math.pow(undefined, 2)` is actually not a crashing error, but it doesn't make much sense, does it?)

All in all, we would prefer developers to:
- either write it like
  ```ts
  Notifications.addNotificationReceivedListener(notification => {
    AsyncStorage.setItem("LAST_POWER", Math.pow((notification.content.data as MyNotificationData).count, 2));
  });
  ```
  which makes it obvious there's a forced type cast here and if the app would crash on notification this would be a good first place to look
- or
  ```ts
  Notifications.addNotificationReceivedListener(notification => {
    const typeSafeData = someParser(typeSchema, notification.content.data);
    if (typeSafeData) {
      AsyncStorage.setItem("LAST_POWER", Math.pow(typeSafeData.count, 2));
    }
  });  
  ```
  where the `someParser` function takes care of validating the contents.

**Pull request closed without merging**
